cmd/proxy-server/main.go

```diff
@@ -540,3 +563,152 @@ func rolesHandler(log *logrus.Entry, opaHost string) http.Handler {
 		defer res.Body.Close()
 	})
 }
+
+func volumesHandler(roleServ *roleClientService, tm token.Manager, log *logrus.Entry) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var sysID, sysType, storPool, tenant string
+		var volumeMap = make(map[string]string)
+		var volumeList []string
+		var resp *pb.RoleListResponse
+		keyTenantRevoked := "tenant:revoked"
+
+		// Verify token
+		log.Info("Verifying token!")
+
+		authz := r.Header.Get("Authorization")
+		parts := strings.Split(authz, " ")
+		if len(parts) != 2 {
+			log.Println("invalid authz header")
+			return
+		}
+		scheme, tkn := parts[0], parts[1]
+
+		switch scheme {
+		case "Bearer":
+			var claims token.Claims
+			//if token can is valid
+			_, err := tm.ParseWithClaims(tkn, JWTSigningSecret, &claims)
+			if err != nil {
+				w.WriteHeader(http.StatusUnauthorized)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			// Check if the tenant is being denied.
+			ok, err := rdb.SIsMember(keyTenantRevoked, claims.Group).Result()
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			if ok {
+				w.WriteHeader(http.StatusUnauthorized)
+				if err := web.JSONErrorResponse(w, fmt.Errorf("tenant is revoked")); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			log.Info("Calling into roles!")
```

Comment:
It would be better to say something about getting/listing roles specifically.

Reply:
fixed, removed log and added more detailed log.